executor/aggregate.go

```diff
@@ -45,39 +96,484 @@ type HashAggExec struct {
+		}
+
+		w := HashAggPartialWorker{
+			baseHashAggWorker: baseAggWorker,
+			inputCh:           make(chan *chunk.Chunk, 1),
+			outputChs:         e.partialOutputChs,
+			giveBackCh:        e.inputCh,
+			globalOutputCh:    e.finalOutputCh,
+			groupSets:         groupSets,
+			aggCtxsMaps:       aggCtxsMappers,
+			groupByItems:      e.GroupByItems,
+		}
+
+		e.partialWorkers[i] = w
+		e.inputCh <- &HashAggInput{
+			chk:        e.children[0].newChunk(),
+			giveBackCh: w.inputCh,
+		}
+	}
+
+	// Init final workers.
+	finalAggFuncs := e.newFinalAggFuncs()
+	for i := 0; i < finalConcurrency; i++ {
+		baseAggWorker := baseHashAggWorker{
+			finishCh:     e.finishCh,
+			aggFuncs:     finalAggFuncs,
+			maxChunkSize: e.maxChunkSize,
+			groupVals:    make([][]byte, 0, 8),
+		}
+		e.finalWorkers[i] = HashAggFinalWorker{
+			baseHashAggWorker:   baseAggWorker,
+			aggCtxsMap:          make(aggCtxsMapper, 0),
+			groupMap:            mvmap.NewMVMap(),
+			inputCh:             e.partialOutputChs[i],
+			outputCh:            e.finalOutputCh,
+			finalResultHolderCh: make(chan *chunk.Chunk, 1),
+			rowBuffer:           make([]types.Datum, 0, e.Schema().Len()),
+			mutableRow:          chunk.MutRowFromTypes(e.retTypes()),
+		}
+		e.finalWorkers[i].finalResultHolderCh <- e.newChunk()
+	}
+}
+
+func (e *HashAggExec) newFinalAggFuncs() (newAggFuncs []aggregation.Aggregation) {
+	newAggFuncs = make([]aggregation.Aggregation, 0, len(e.AggFuncs))
+	idx := 0
+	for _, af := range e.AggFuncs {
+		var aggFunc aggregation.Aggregation
+		idx, aggFunc = af.GetFinalAggFunc(idx)
+		newAggFuncs = append(newAggFuncs, aggFunc)
+	}
+	return newAggFuncs
+}
+
+// HashAggPartialWorker gets and handles origin data or partial data from inputCh,
+// then shuffle the intermediate results to corresponded final workers.
+func (w *HashAggPartialWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
+	var (
+		chk         *chunk.Chunk
+		ok          bool
+		sc          = ctx.GetSessionVars().StmtCtx
+		needShuffle = false
+	)
+	defer func() {
+		if needShuffle {
+			w.shuffleIntermData(sc)
+		}
+		waitGroup.Done()
+	}()
+	for {
+		select {
+		case <-w.finishCh:
+			return
+		case chk, ok = <-w.inputCh:
+			if !ok {
+				return
+			}
+		}
+
+		if !w.updateIntermData(ctx, sc, chk, len(w.aggCtxsMaps)) {
+			return
+		}
+		needShuffle = true
+		if chk.NumRows() == 0 {
+			return
+		}
+		w.giveBackCh <- &HashAggInput{
+			chk:        chk,
+			giveBackCh: w.inputCh,
+		}
+	}
+}
+
+func (w *HashAggPartialWorker) updateIntermData(ctx sessionctx.Context, sc *stmtctx.StatementContext, chk *chunk.Chunk, finalConcurrency int) (isOk bool) {
+	inputIter := chunk.NewIterator4Chunk(chk)
+	for row := inputIter.Begin(); row != inputIter.End(); row = inputIter.Next() {
+		groupKey, err := w.getGroupKey(sc, row)
+		if err != nil {
+			w.globalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return false
+		}
+		finalWorkerIdx := int(murmur3.Sum32(groupKey)) % finalConcurrency
```

Comment:
Do not calculate `finalWorkerIdx` for every group, we can calculate it when we need to shuffle the partial result of a group to the final workers. We don't need to maintain `aggCtxsMaps`, `groupSets` for every `finalWorkerIdx` in the partial workers, which can help us to simplify the complexity of partial workers, and we can avoid the `if` branch in line 330.